zkvm/spec/ZkVM.md

```diff
@@ -1957,9 +1969,100 @@ Now let’s see how [`borrow`](#borrow) simplifies things: it allows you to make
+
+One way to do that is to break the scalar in 253 bits (using 253 multipliers) and add a disjunction "at least one of these bits is 1" (using additional 252 multipliers), spending total 505 multipliers (this is 8x more expensive than a regular 64-bit rangeproof).
+
+On the other hand, using `not` may not be a good way to express contracts because of a dramatic increase in complexity: a contract that says `not(B)` effectively inverts a small set of inadmissible inputs, producing a big set of addmissible inputs.
+
+We need to investigate whether there are use-cases that cannot be safely or efficiently expressed with only [`and`](#and) and [`or`](#or) combinators.
+
+
+### What ensures transaction uniqueness?
+
+In ZkVM:
+
+* [Transaction ID](#transaction-id) is globally unique,
+* [UTXO ID](#utxo) is globally unique,
+* [Nonce](#nonce) is globally unique,
+* [Value](#value-type) is not unique,
+* [Contract](#contract-type) is not unique.
+
+**Rationale:**
+
+TxVM ensures transaction uniqueness this way:
+
+* Each value has a unique “anchor” (32-byte payload)
+* When values are split/merged, anchor is one-way hashed to produce new anchor(s).
+* `finalize` instruction consumes a zero-quantity value, effectively consuming a unique anchor.
+* Anchors are originally produced via `nonce` instruction that uses blockchain state to prevent reuse of nonces.
+* Issuance of new assets consumes a zero-quantity value, moving the anchor from it to the new value.
+
+**Pro:**
+
+UTXO ID is fully determined before transaction is finalized. So e.g. a child transaction can be formed before the current transaction is completed and its ID is known. This might be handy in some cases.
+
+**Meh:**
+
+It seems like transaction ID is defined quite simply as a hash of the log, but it also must consume an anchor (since UTXOs alone are not guaranteed to be unique, and one can make a transaction without spending any UTXOs). So transaction ID computation still has some level of special handling.
+
+**Con:**
+
+Recipient of payment cannot fully specify the contract snapshot because they do not know sender’s anchors. This is not a problem in cleartext TxVM, but a problem in ZkVM where values have to participate in the Cloak protocol.
+
+Storing anchor inside a value turned out to be handy, but is not very ergonomic. For instance, a contract cannot simply “claim” an arbitrary value and produce a negative counterpart, w/o having _some_ value to “fork off” an anchor from.
+
+Another potential issue: UTXOs are not guaranteed to be always unique. E.g. if a contract does not modify its value and other content, it can re-save itself to the same UTXO ID. It can even toggle between different states, returning to the previously spent ID. This can cause issues in some applications that forget that UTXO ID in special circumstances can be resurrected.
```

Comment:
> It can even toggle between different states

This is reminiscent of [the "ko" rule in Go](https://en.wikipedia.org/wiki/Go_(game)#Ko_rule):
> Players are not allowed to make a move that returns the game to the previous position.

Reply:
Nice. Seems like "Ko rule" prohibits immediate return to previous position, but it's more complex if you come to the previous state after N moves. ZkVM prohibits returning to the previous UTXO ID within N < 2^64 moves (collisions get probable at around N=2^128 steps).